contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L2DepositedERC20.sol

```diff
@@ -0,0 +1,159 @@
+    )
+        public
+    {
+        require(address(l1ERC20Gateway) == address(0), "Contract has already been initialized");
+
+        l1ERC20Gateway = _l1ERC20Gateway;
+        
+        emit Initialized(l1ERC20Gateway);
+    }
+
+    /**********************
+     * Function Modifiers *
+     **********************/
+
+    modifier onlyInitialized() {
+        require(address(l1ERC20Gateway) != address(0), "Contract has not yet been initialized");
+        _;
+    }
+
+    /***************
+     * Withdrawing *
+     ***************/
+
+    /**
+     * @dev initiate a withdraw of some ERC20 to the caller's account on L1
+     * @param _amount Amount of the ERC20 to withdraw
+     */
+    function withdraw(
+        uint _amount
+    )
+        external
+        override
+        onlyInitialized()
+    {
+        _initiateWithdrawal(msg.sender, _amount);
+    }
+
+    /**
+     * @dev initiate a withdraw of some ERC20 to a recipient's account on L1
+     * @param _to L1 adress to credit the withdrawal to
+     * @param _amount Amount of the ERC20 to withdraw
+     */
+    function withdrawTo(address _to, uint _amount) external override onlyInitialized() {
+        _initiateWithdrawal(_to, _amount);
+    }
+
+    /**
+     * @dev Performs the logic for deposits by storing the ERC20 and informing the L2 ERC20 Gateway of the deposit.
+     *
+     * @param _to Account to give the withdrawal to on L1
+     * @param _amount Amount of the ERC20 to withdraw
+     */
+    function _initiateWithdrawal(address _to, uint _amount) internal {
+        // burn L2 funds so they can't be used more on L2
+        _burn(msg.sender, _amount);
+
+        // Construct calldata for l1ERC20Gateway.finalizeWithdrawal(_to, _amount)
+        bytes memory data = abi.encodeWithSelector(
+            iOVM_L1ERC20Gateway.finalizeWithdrawal.selector,
+            _to,
+            _amount
+        );
+
+        // Send message up to L1 gateway
+        sendCrossDomainMessage(
+            address(l1ERC20Gateway),
+            data,
+            8999999 // TODO: meter and set with some buffer (actually I think its currently unused in this direction)
+        );
+
+        emit WithdrawalInitiated(msg.sender, _to, _amount);
+    }
+
+    /************************************
+     * Cross-chain Function: Depositing *
+     ************************************/
+
+    /**
+     * @dev Complete a deposit from L1 to L2, and credits funds to the recipient's balance of this 
+     * L2 ERC20 token. 
+     * This call will fail if it did not originate from a corresponding deposit in OVM_L1ERC20Gateway. 
+     *
+     * @param _to Address to receive the withdrawal at
+     * @param _amount Amount of the ERC20 to withdraw
+     */
+    function finalizeDeposit(address _to, uint _amount) external override onlyInitialized()
+        onlyFromCrossDomainAccount(address(l1ERC20Gateway))
+    {
+        _mint(_to, _amount);
+        emit DepositFinalized(_to, _amount);
+    }
+}
```

Comment:
Nit: There are a bunch of functions in this file with non-standard styling. -- finalizeDeposit, _initiateWithdrawal, withdrawTo